src/main/java/org/truffleruby/parser/BodyTranslator.java

```diff
@@ -1695,86 +1702,98 @@ public RubyNode visitGlobalAsgnNode(GlobalAsgnParseNode node) {
                 RubyNode processClass = new ObjectLiteralNode(context.getCoreLibrary().getProcessModule());
-                translated = new RubyCallNode(new RubyCallNodeParameters(processClass, "setproctitle", null,
-                                new RubyNode[]{writeGlobalVariableNode}, false, false));
-                translated.unsafeSetSourceSection(sourceSection);
+                ret = new RubyCallNode(new RubyCallNodeParameters(processClass, "setproctitle", null,
+                        new RubyNode[]{ writeGlobalVariableNode }, false, false));
             } else {
-                translated = writeGlobalVariableNode;
+                ret = writeGlobalVariableNode;
             }
-
-            return addNewlineIfNeeded(node, translated);
         }
+
+        return addNewlineIfNeeded(node, withSourceSection(sourceSection, ret));
     }
 
     @Override
     public RubyNode visitGlobalVarNode(GlobalVarParseNode node) {
-        String name = node.getName();
+        final String variableName = node.getName();
 
-        if (GLOBAL_VARIABLE_ALIASES.containsKey(name)) {
-            name = GLOBAL_VARIABLE_ALIASES.get(name);
-        }
+        return new LazyRubyNode(() -> readGlobal(node, variableName));
+    }
 
+    private RubyNode readGlobal(ParseNode node, String variableName) {
         final SourceIndexLength sourceSection = node.getPosition();
-        final RubyNode ret;
+        String name = variableName;
 
-        if (FRAME_LOCAL_GLOBAL_VARIABLES.contains(name)) {
-            // Assignment is implicit for many of these, so we need to declare when we use
+        if (context != null) {
+            name = context.getCoreLibrary().getGlobalVariables().getOriginalName(name);
+        }
 
-            RubyNode readNode = environment.findFrameLocalGlobalVarNode(name, source, sourceSection);
 
-            if (THREAD_AND_FRAME_LOCAL_GLOBAL_VARIABLES.contains(name)) {
-                readNode = new GetFromThreadAndFrameLocalStorageNode(readNode);
-                readNode.unsafeSetSourceSection(sourceSection);
+        final RubyNode ret;
+        if (GlobalVariables.BACKREF_GLOBAL_VARIABLES.contains(name)) {
+            int index = 0;
+            final char type = name.charAt(1);
+            switch (type) {
+                case '`':
+                    index = ReadMatchReferenceNode.PRE;
+                    break;
+                case '\'':
+                    index = ReadMatchReferenceNode.POST;
+                    break;
+                case '&':
+                    index = ReadMatchReferenceNode.GLOBAL;
+                    break;
+                case '+':
+                    index = ReadMatchReferenceNode.HIGHEST;
+                    break;
+                default:
+                    if ('1' <= type && type <= '9') {
+                        index = (type - '0');
+                    } else {
+                        throw new UnsupportedOperationException(name);
```

Comment:
Minor: But it'd be nicer to have a more descriptive message here.

Reply:
It was already like that and anyway it's just unreachable.
Feel free to improve it if you care though.

Reply:
I'll just remove the branch and make it an assert, that's more explicit.

Reply:
Our use of `UnsupportedOperationException` with no or very terse messages really bothers me. Just changing to `"unknown global: " + name` would be a big help. But, since it was already like this we can just leave it.

Reply:
For impossible cases like this, I think it really does not matter, just the Java line is enough, and then maybe the name in case of spurious failures.

Reply:
I could just be dense, but nothing about this tells me that this is obviously an impossible case. I can't tell from the context. There's no comment. The exception type doesn't indicate it. The exception message doesn't indicate it either. You really have to read this code to figure out it's an impossible case and have context about the permissible global variable names.

If I'm alone in that, I'll drop it. But generally it costs nothing to have a more descriptive message to help with debugging.

Reply:
But even a message like `unknown backref global` would not tell much without looking at the code and how this can happen. And then since this logic is guarded by the name being in the Set and we handle all possible cases, the only possibility would be a VM bug, in which case the message will be of little help. But I'll add a simple message to the `assert`.